sources/api/migration/migrations/v1.15.0/oci-defaults-max-open-files/src/main.rs

```diff
@@ -0,0 +1,68 @@
+use migration_helpers::common_migrations::NoOpMigration;
+use migration_helpers::{migrate, Migration, MigrationData, Result};
+use serde_json::Value;
+use std::process;
+
+const HARD_RESOURCE_LIMIT_SETTING_NAME: &str =
+    "settings.oci-defaults.resource-limits.max-open-files.hard-limit";
+const SOFT_RESOURCE_LIMIT_SETTING_NAME: &str =
+    "settings.oci-defaults.resource-limits.max-open-files.soft-limit";
+
+/// This migration changes the hard and soft limit for rlimit_nofile to u32 from i64 on downgrade.
+/// There is no need of migration on upgrade as u32 will automatically change to i64
+pub struct ChangeMaxOpenFileResourceLimitType;
+
+fn convert_to_u_32(value: &mut Value) {
+    let s = if value.is_i64() {
+        let v: i64 = serde_json::from_value(value.clone()).unwrap();
+        match v {
+            -1 => u32::MAX,
+            v if v > u32::MAX as i64 => u32::MAX,
+            _ => v as u32,
+        }
+    } else {
+        return;
+    };
+    *value = Value::Number(s.into());
+}
+
+impl Migration for ChangeMaxOpenFileResourceLimitType {
+    /// On upgrade there is nothing to do (see above).
+    fn forward(&mut self, input: MigrationData) -> Result<MigrationData> {
+        Ok(input)
```

Comment:
If we did an end to end migration that was set as `unlimited`, would we end up with `unlimited` again or `u32::MAX`? This would be pretty surprising if its not getting us back to `unlimited`.

Reply:
It feels like `u32::MAX` is the best we can do in that case - earlier versions didn't support anything beyond that, or "unlimited". Removing the limit entirely might cause it to be set much lower than expected, and `u32::MAX` open files is still quite a few.

Reply:
Yeah, I think I'm ok with `u32::MAX` over dropping it. I was considering if we should take `u32::MAX` on the forward case as `unlimited` but that might be less desirable than leaving it as we find it. Either way, I think the delta between `u32::MAX` and `unlimited` actually mattering in the real world is pretty low likelihood.